versioned_hdf5/hyperspace.py

```diff
@@ -0,0 +1,366 @@
+    shape: Py_ssize_t[:],
+    out_a: Py_ssize_t[:, :],
+    out_b: Py_ssize_t[:, :],
+    cur: Py_ssize_t,
+) -> Py_ssize_t:
+    """Helper of fill_hyperspace.
+
+    Given two ordered points in a N-dimensional hyperspace, yield the top-left
+    corner (included) and bottom-right corner (excluded) of the N-dimensional
+    hyperrectangles that cover the whole space between the two points (both excluded).
+
+    In other words, fill the flattened vector in the range ]obstacle1, obstacle2[.
+
+    Parameters
+    ----------
+    obstacle1:
+        The first point to avoid in the N-dimensional hyperspace
+    obstacle2:
+        The second point to avoid in the N-dimensional hyperspace.
+        It must be after the first point along the flattened vector.
+    shape:
+        The shape of the whole N-dimensional hyperspace to cover
+    out_a:
+        Empty buffer to write the top-left corners into
+    out_b:
+        Empty buffer to write the bottom-right corners into
+    cur:
+        row index of out_a and out_b to start writing from.
+        out_a and out_b must be at least 2**ndim - 1 + cur rows in size.
+
+    Returns
+    -------
+        row index of out_a and out_b after the one that was last written to
+
+    Example:
+    --------
+    Given two points at coordinates (3, 4, 5) and (3, 7, 8)
+    in a space of shape (10, 10, 10), this function yields
+
+    - [ 3,   4, 6:] -> (3, 4, 6), (4, 5, 10)
+    - [ 3, 5:7,  :] -> (3, 5, 0), (4, 7, 10)
+    - [ 3,   7, :8] -> (3, 7, 0), (4, 8,  8)
+
+    >>> out = np.empty((2**3 - 1, 6), dtype=np.intp)
+    >>> _hyperrectangles_between(
+    ...     np.array([3, 4, 5]),
+    ...     np.array([3, 7, 8]),
+    ...     np.array([10, 10, 10]),
+    ...     out[:, :3],
+    ...     out[:, 3:],
+    ...     0
+    ... )
+    3
+    >>> out[:3]
+    array([[ 3,  4,  6,  4,  5, 10],
+           [ 3,  5,  0,  4,  7, 10],
+           [ 3,  7,  0,  4,  8,  8]])
+    """
+    ndim = len(obstacle1)  # Always >0
+
+    if ndim == 1:
+        # Trivial use case in mono-dimensional space.
+        # There can be at most one 1D line between the two obstacles.
+        #
+        # ..XaaX...
+
+        i, j = obstacle1[0], obstacle2[0]
+        if i + 1 < j:
+            out_a[cur, 0] = i + 1
+            out_b[cur, 0] = j
+            cur += 1
+        return cur
+
+    if obstacle1[0] == obstacle2[0]:
+        # The two obstacles are on the same outermost hyperspace/space/surface/row.
+        # Simplify the problem by recursing into a hyperspace with one less dimension.
+        #
+        # .........
+        # ..XaaX...
+        # .........
+
+        new_cur = _hyperrectangles_between(
+            obstacle1[1:],
+            obstacle2[1:],
+            shape[1:],
+            out_a[:, 1:],
+            out_b[:, 1:],
+            cur,
+        )
+        prefix = obstacle1[0]
+        out_a[cur:new_cur, 0] = prefix
+        out_b[cur:new_cur, 0] = prefix + 1
+        return new_cur
+
```

Comment:
What comes below cost me two solid days of headache trying to visualize in my head spaces with more than 3 dimensions